pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/SegmentPreprocessThrottler.java

```diff
@@ -0,0 +1,226 @@
+import com.google.common.base.Preconditions;
+import java.util.Map;
+import java.util.Set;
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.pinot.common.concurrency.AdjustableSemaphore;
+import org.apache.pinot.spi.config.provider.PinotClusterConfigChangeListener;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Used to throttle the total concurrent index rebuilds that can happen on a given Pinot server.
+ * Code paths that do no need to rebuild the index or which don't happen on the server need not utilize this throttler.
+ */
+public class SegmentPreprocessThrottler implements PinotClusterConfigChangeListener {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentPreprocessThrottler.class);
+
+  /**
+   * _maxPreprocessConcurrency and _maxConcurrentPreprocessesBeforeServingQueries must be >= 0. To effectively disable
+   * throttling, this can be set to a very high value
+   */
+  private int _maxPreprocessConcurrency;
+  private int _maxPreprocessConcurrencyBeforeServingQueries;
+  private boolean _isServingQueries;
+  private final AdjustableSemaphore _semaphore;
+
+  /**
+   * @param maxPreprocessConcurrency configured preprocessing concurrency
+   * @param maxPreprocessConcurrencyBeforeServingQueries configured preprocessing concurrency before serving queries
+   * @param isServingQueries whether the server is ready to serve queries or not
+   */
+  public SegmentPreprocessThrottler(int maxPreprocessConcurrency, int maxPreprocessConcurrencyBeforeServingQueries,
+      boolean isServingQueries) {
+    LOGGER.info("Initializing SegmentPreprocessThrottler, maxPreprocessConcurrency: {}, "
+            + "maxPreprocessConcurrencyBeforeServingQueries: {}, isServingQueries: {}",
+        maxPreprocessConcurrency, maxPreprocessConcurrencyBeforeServingQueries, isServingQueries);
+    Preconditions.checkArgument(maxPreprocessConcurrency > 0,
+        "Max preprocess parallelism must be > 0, but found to be: " + maxPreprocessConcurrency);
+    Preconditions.checkArgument(maxPreprocessConcurrencyBeforeServingQueries > 0,
+        "Max preprocess parallelism before serving queries must be > 0, but found to be: "
+            + maxPreprocessConcurrencyBeforeServingQueries);
+
+    _maxPreprocessConcurrency = maxPreprocessConcurrency;
+    _maxPreprocessConcurrencyBeforeServingQueries = maxPreprocessConcurrencyBeforeServingQueries;
+    _isServingQueries = isServingQueries;
+
+    // maxConcurrentPreprocessesBeforeServingQueries is only used prior to serving queries and once the server is
+    // ready to serve queries this is not used again. Thus, it is safe to only pick up this configuration during
+    // server startup. There is no need to allow updates to this via the ZK CLUSTER config handler
+    int preServeQueryParallelism = Math.max(_maxPreprocessConcurrency, _maxPreprocessConcurrencyBeforeServingQueries);
```

Comment:
We probably don't want this max logic. It doesn't align with logic in `handleMaxPreprocessConcurrencyBeforeServingQueriesChange()`

Reply:
oh yes, good catch, I forgot this clean up